src/ray/common/asio/asio_util.h

```diff
@@ -82,4 +88,98 @@ class InstrumentedIOContextWithThread {
   instrumented_io_context io_service_;
   boost::asio::io_service::work work_;  // to keep io_service_ running
   std::thread io_thread_;
+  std::string name_;
+};
+
+/// `IoContextProvider` uses a specified `Policy` to determine whether a type `T`
+/// requires a dedicated `io_context` or should use a shared default `io_context`.
+/// It provides a method to retrieve the appropriate `io_context` for instances of
+/// different classes.
+///
+/// @param Policy The policy class that defines which types require dedicated
+/// `io_context` instances.
+///
+/// ## The Policy
+/// SYNOPSIS:
+/// ```
+/// struct YourPolicy {
+///     // List of all IO Context names. We will create 1 thread + 1
+///     // instrumented_io_context for each. Must be unique and should not contain empty
+///     // names.
+///     constexpr static std::array<std::string_view, N> kAllDedicatedIoContextNames;
+///
+///     // For a given T, returns an index to kAllDedicatedIoContextNames, or -1 for the
+///     // default io context.
+///     constexpr static std::string_view GetDedicatedIoContextIndex<T>();
+/// }
+/// ```
+///
+/// For an example, see `GcsServerIoContextPolicy`.
+///
+/// ## Notes
+///
+/// - `default_io_context` must outlive the `IoContextProvider` instance.
+/// - Eagerly creates dedicated `io_context` instances in ctor.
+/// - NOT thread-safe. Please always access to this class from the same thread.
+/// - There is no way to remove a dedicated `io_context` once created until destruction.
+template <typename Policy>
+class IoContextProvider {
```

Comment:
```suggestion
class IOContextProvider {
```

everywhere for consistency.